src/ipfsd-client.js

```diff
@@ -4,108 +4,88 @@ const request = require('superagent')
 class Client {
-  constructor (baseUrl, _id, initialized, apiAddr, gwAddrs, options) {
-    this.options = options || {}
+  constructor (baseUrl, remoteState, options = {}) {
+    this.options = options
     this.baseUrl = baseUrl
-    this._id = _id
-    this._apiAddr = multiaddr(apiAddr)
-    this._gwAddr = multiaddr(gwAddrs)
-    this.initialized = initialized
-    this.started = false
-    this.api = createApi(apiAddr, gwAddrs, this.options.IpfsApi || IpfsApi)
-  }
-
-  /**
-   * Get the address of connected IPFS API.
-   *
-   * @returns {Multiaddr}
-   */
-  get apiAddr () {
-    return this._apiAddr
+    this._id = remoteState._id
+    this.initialized = remoteState.initialized
+    this.started = remoteState.started
+    this.clean = true
+    this.apiAddr = null
+    this.gatewayAddr = null
+    this.path = remoteState.path
+    this.api = null
+
+    if (this.started) {
+      this.setApi(remoteState.apiAddr)
+      this.setGateway(remoteState.gatewayAddr)
+    }
   }
 
-  /**
-   * Set the address of connected IPFS API.
-   *
-   * @param {Multiaddr} addr
-   * @returns {void}
-   */
-  set apiAddr (addr) {
-    this._apiAddr = addr
+  setApi (addr) {
+    if (addr) {
+      this.apiAddr = multiaddr(addr)
+      this.api = (this.options.IpfsApi || IpfsApi)(addr)
+      this.api.apiHost = this.apiAddr.nodeAddress().address
+      this.api.apiPort = this.apiAddr.nodeAddress().port
+    }
   }
 
-  /**
-   * Get the address of connected IPFS HTTP Gateway.
-   *
-   * @returns {Multiaddr}
-   */
-  get gatewayAddr () {
-    return this._gwAddr
+  setGateway (addr) {
+    if (addr) {
+      this.gatewayAddr = multiaddr(addr)
+      this.api.gatewayHost = this.gatewayAddr.nodeAddress().address
+      this.api.gatewayPort = this.gatewayAddr.nodeAddress().port
+    }
   }
 
   /**
-   * Set the address of connected IPFS Gateway.
+   * Initialize a repo.
    *
-   * @param {Multiaddr} addr
+   * @param {Object} [initOptions={}]
+   * @param {number} [initOptions.keysize=2048] - The bit size of the identiy key.
+   * @param {string} [initOptions.directory=IPFS_PATH] - The location of the repo.
+   * @param {function (Error, Node)} callback
    * @returns {void}
    */
-  set gatewayAddr (addr) {
-    this._gwAddr = addr
-  }
+  init (initOptions, callback) {
+    if (typeof initOptions === 'function') {
+      callback = initOptions
+      initOptions = null
+    }
 
-  /**
-   * Initialize a repo.
-   *
-   * @param {Object} [initOpts={}]
-   * @param {number} [initOpts.keysize=2048] - The bit size of the identiy key.
-   * @param {string} [initOpts.directory=IPFS_PATH] - The location of the repo.
-   * @param {function (Error, Node)} cb
-   * @returns {undefined}
-   */
-  init (initOpts, cb) {
-    if (typeof initOpts === 'function') {
-      cb = initOpts
-      initOpts = {}
+    if (this.initialized && initOptions) {
+      return callback(new Error(`Repo already initialized can't use different options, ${JSON.stringify(initOptions)}`))
```

Comment:
We are going in a way of having all the errors using [err-code](https://www.npmjs.com/package/err-code). I know that this is probably one of the repos where it is not that important to have the codes, but since we are creating new ones, I think we can add it here.

Reply:
Same happens in the 2 other daemons.

Reply:
one of my okrs is to add error codes everywhere. i would prefer to handle that in another PR